src/main/scala/org/zalando/kanadi/api/Subscriptions.scala

```diff
@@ -524,13 +542,197 @@ final case class CancelledByClient(subscriptionId: SubscriptionId, streamId: Str
     s"Stream cancelled by client SubscriptionId: ${subscriptionId.id}, StreamId: ${streamId.id}"
 }
 
-case class Subscriptions(baseUri: URI, oAuth2TokenProvider: Option[OAuth2TokenProvider] = None)(implicit
+object AvroSubscriptions {
+
+  def apply[T](baseUri: URI,
+               oAuth2TokenProvider: Option[OAuth2TokenProvider],
+               eventTypeName: EventTypeName,
+               consumerSchema: String,
+               eventTypes: EventTypesInterface)(implicit
+      kanadiHttpConfig: HttpConfig,
+      http: HttpExt,
+      materializer: Materializer,
+      executionContext: ExecutionContext,
+      tag: ClassTag[T]): AvroSubscriptions[T] = {
+    implicit val exponentialBackoffConfig = ExponentialBackoffConfig(FiniteDuration(100, TimeUnit.MILLISECONDS), 1.5, 2)
+    val result = eventTypes
+      .fetchMatchingSchema(eventTypeName, consumerSchema)
+      .map(etSchemaOpt =>
+        etSchemaOpt.map(etSchema =>
+          new AvroSubscriptions[T](baseUri,
+                                   oAuth2TokenProvider,
+                                   AvroSchema(eventTypeName, etSchema.schema.asString.get, etSchema.version.get))))
+
+    Await.result(result, Duration.apply(5, TimeUnit.SECONDS)).getOrElse(throw SchemaNotFoundError(consumerSchema))
+  }
+}
+class AvroSubscriptions[T](baseUri: URI,
+                           oAuth2TokenProvider: Option[OAuth2TokenProvider] = None,
+                           consumerSchema: AvroSchema)(implicit
+    kanadiHttpConfig: HttpConfig,
+    http: HttpExt,
+    materializer: Materializer,
+    tag: ClassTag[T])
+    extends Subscriptions(baseUri, oAuth2TokenProvider) {
+
+  import com.fasterxml.jackson.dataformat.avro.{AvroSchema => JacksonSchema}
+
+  private val userSchema = new JacksonSchema(consumerSchema.parsedSchema)
+  private val reader     = AvroUtil.AvroMapper.reader(userSchema)
+
+  def eventsStreamedSourceAvro(subscriptionId: SubscriptionId,
```

Comment:
> def eventsStreamedSourceAvro

IMHO the method is too big and might be difficult to follow, it might be worth so split it into several methods that are more 'granular'.